libs/small-steps/src/Control/State/Transition/Extended.hs

```diff
@@ -341,26 +348,28 @@ validateTrans ::
   (e -> PredicateFailure sts) ->
   Validation (NonEmpty e) () ->
   Rule sts ctx ()
-validateTrans t v = liftF $ Predicate [] v t ()
+validateTrans t v = liftF $ Predicate v t ()
 
 -- | Same as `validation`, except with ability to translate opaque failures
 -- into `PredicateFailure`s with a help of supplied function.
 validateTransLabeled ::
   -- | Transformation function for all errors
   (e -> PredicateFailure sts) ->
   -- | Supply a list of labels to be used as filters when STS is executed
-  [Label] ->
+  NonEmpty Label ->
   -- | Actual validations to be executed
   Validation (NonEmpty e) () ->
   Rule sts ctx ()
-validateTransLabeled t labels v = liftF $ Predicate labels v t ()
+validateTransLabeled t labels v = liftF $ Label labels (liftF $ Predicate v t ()) ()
 
 -- | Oh noes!
 --
 --   This takes a condition (a boolean expression) and a failure and results in
 --   a clause which will throw that failure if the condition fails.
 (?!) :: Bool -> PredicateFailure sts -> Rule sts ctx ()
-(?!) = labeledPred []
+(?!) cond onFail =
+  liftF $
+    Predicate (if cond then Success () else Failure (() :| [])) (const onFail) ()
```

Comment:
Just one minor optional suggestion, which doesn't change the semantics, but might make it easier to read the code:
```suggestion
    Predicate (if cond then Success () else Failure (onFail :| [])) id ()
```